### PR TITLE
Possible solution to making Semsiman object immutable.

### PR DESCRIPTION
**Pull request by hrshdhgd**
Using `Mutex` 

Addresses https://github.com/monarch-initiative/semsimian/issues/112

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -42,8 +43,19 @@ pub fn get_rss_instance() -> RustSemsimian {
     ]);
 
     let assoc_predicate: HashSet<TermID> = HashSet::from(["biolink:has_phenotype".to_string()]);
+    let rss = Mutex::new(Some(RustSemsimian::new(None, predicates, None, db)));
 
-    let mut rss = RustSemsimian::new(None, predicates, None, db);
-    rss.pregenerate_cache(&assoc_predicate, &SearchTypeEnum::Flat);
-    rss
+    {
```
Comment (caufieldjh) [inline src/utils.rs:51]: this part (lines 48 onward) would be most useful as its own utility function in Semsimian, along with line 46 above, along with a parameter for passing a specific predicate or predicate list (and the whole target DB, for that matter)

Comment (hrshdhgd) [reply src/utils.rs:51]: Agreed. Once confirmed that this solution works and does not have unforeseen side-effects, we can implement this in `semsimian` thus making this just a simple function call.

**Comment by hrshdhgd** [comment]
An alternative solution could be from the rocket side itself:
```
rocket::ignite()
        .manage(Mutex::new(Semsimian { /* initialize fields here */ }))
        .mount("/", routes![index])
        .launch();
```

Not tested.

As per ChatGPT:
```
In this example, `Semsimian` is wrapped in a Mutex, which ensures that only one worker can access it at a time. When a worker wants to access `Semsimian`, it must first acquire the lock. If another worker is currently holding the lock, it will have to wait until the lock is released.

Keep in mind that while sharing a single object across all workers can be useful in certain scenarios, it can also lead to increased contention and decreased performance if not used carefully. It's important to consider the specific needs of your application when deciding whether to use this approach.
```

This caution applies to this PR solution as well.

**Comment by glass-ships** [comment]
apologies i still haven't gotten a chance to test this yet. 

honestly the main hold up for me is that i'm not actually sure how exactly to use the `full_search()` function (ie, what the inputs for each argument should actually be)

i'd planned on waiting until kevin returned from vacation to work with him on testing it out, but if anyone has some suggestions for an example usage i can work with that to make sure it at least works?

**Comment by hrshdhgd** [comment]
Not a problem. Here's a test in semsimian to run full search:



```rust
    fn test_associations_search_phenio_mondo() {
        let mut db_path = PathBuf::new();
        if let Some(home) = std::env::var_os("HOME") {
            db_path.push(home);
            db_path.push(".data/oaklib/phenio.db");
        } else {
            panic!("Failed to get home directory");
        }
        let db = Some(db_path.to_str().expect("Failed to convert path to string"));
        let predicates: Option<Vec<Predicate>> = Some(vec![
            "rdfs:subClassOf".to_string(),
            "BFO:0000050".to_string(),
        ]);

        let mut rss = RustSemsimian::new(None, predicates, None, db);

        rss.update_closure_and_ic_map();

        let assoc_predicate: HashSet<TermID> = HashSet::from(["biolink:has_phenotype".to_string()]);
        let subject_prefixes: Option<Vec<TermID>> = Some(vec!["MONDO:".to_string()]);
        let object_terms: HashSet<TermID> = HashSet::from([
            "HP:0008132".to_string(),
            "HP:0000189".to_string(),
            "HP:0000275".to_string(),
            "HP:0000276".to_string(),
            "HP:0000278".to_string(),
            "HP:0000347".to_string(),
            "HP:0001371".to_string(),
            "HP:0000501".to_string(),
            "HP:0000541".to_string(),
            "HP:0000098".to_string(),
        ]);
        let search_type: SearchTypeEnum = SearchTypeEnum::Full;
        let limit: Option<usize> = Some(20);

        // Call the function under test
        let result = rss.associations_search(
            &assoc_predicate,
            &object_terms,
            true,
            &None,
            &subject_prefixes,
            &search_type,
            limit,
        );
        let unique_scores: HashSet<_> =
            result.iter().map(|(score, _, _)| score.to_bits()).collect();
        let count = unique_scores.len();
        assert!(count <= limit.unwrap());
    }
```

Is it good enough for a starting point?

**Comment by hrshdhgd** [comment]
~Remember here we're using the `go-nucleus.db` as a resource. The resource, subject and object sets will change for you accordingly.~

Updated to reflect MONDO

**Comment by glass-ships** [comment]
I have yet to test `full_search()` with this PR (hoping to get to it soon, apologies!) 

However, I should note that compare_termsets no longer seems to work and it's not immediately clear to me why: 
```
thread 'rocket-worker-thread' panicked at 'Resource returned nothing with predicates: no such table: entailed_edge', /home/glass/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.11/src/lib.rs:99:29
...
   >> Handler compare_termsets panicked.
   >> This is an application bug.
   >> A panic in Rust must be treated as an exceptional event.
   >> Panicking is not a suitable error handling mechanism.
   >> Unwinding, the result of a panic, is an expensive operation.
   >> Panics will degrade application performance.
   >> Instead of panicking, return `Option` and/or `Result`.
   >> Values of either type can be returned directly from handlers.
   >> A panic is treated as an internal server error.
   >> Outcome: Error(500 Internal Server Error)
   >> No 500 catcher registered. Using Rocket default.
   >> Response succeeded.
```

This seems to happen on main too actually, perhaps i should make a separate issue? 
I think the main reason we don't see this issue affecting the live website is we haven't done a new deployment since some of the most recent changes to this repo

**Comment by glass-ships** [comment]
looks like the above happens when trying to call the `test_associations_search_phenio_mondo()` function as well

**Comment by hrshdhgd** [comment]
> Resource returned nothing with predicates: no such table: entailed_edge

- You may have a corrupted phenio.db file.
- If you've updated the resource, check if it has all tables the previous resource had.

**Comment by glass-ships** [comment]
ah yep, redownloaded the database and that seemed to fix it! 

made a simple endpoint to test your function which also seems to work:
```
#[get("/search")]
pub fn full_search() {
    test_associations_search_phenio_mondo();
}
```
-> 
```
GET /search text/html:
   >> Matched: (full_search) GET /search
[00:00:20] Building closure and IC map: ████████████████████████████████████████ 100%                                                                                                              Cache "MONDO:biolink:has_phenotypefull" not yet generated...
Generating cache! "MONDO:biolink:has_phenotypefull"
   >> Outcome: Success(200 OK)
   >> Response succeeded.
```

so it at least seems like this should work for our use case!

**Comment by glass-ships** [comment]
Actually, I modified the test to try and re-use the same RSS that's defined higher up in `lib.rs`, but this _doesn't_ seem to work as well as when the test function creates is own instance of rust semsimian:

modified code:
```
fn test_associations_search_phenio_mondo() {
    let assoc_predicate: HashSet<TermID> = HashSet::from(["biolink:has_phenotype".to_string()]);
    let subject_prefixes: Option<Vec<TermID>> = Some(vec!["MONDO:".to_string()]);
    let object_terms: HashSet<TermID> = HashSet::from([
        "HP:0008132".to_string(),
        "HP:0000189".to_string(),
        "HP:0000275".to_string(),
        "HP:0000276".to_string(),
        "HP:0000278".to_string(),
        "HP:0000347".to_string(),
        "HP:0001371".to_string(),
        "HP:0000501".to_string(),
        "HP:0000541".to_string(),
        "HP:0000098".to_string(),
    ]);
    let search_type: SearchTypeEnum = SearchTypeEnum::Full;
    let limit: Option<usize> = Some(20);

    // Call the function under test
    let result = RSS.associations_search(
        &assoc_predicate,
        &object_terms,
        true,
        &None,
        &subject_prefixes,
        &search_type,
        limit,
    );
    let unique_scores: HashSet<_> = result.iter().map(|(score, _, _)| score.to_bits()).collect();
    let count = unique_scores.len();
    assert!(count <= limit.unwrap());
}
```

Build error:
```
error[E0596]: cannot borrow data in dereference of `RSS` as mutable
   --> src/lib.rs:97:18
    |
97  |       let result = RSS.associations_search(
    |  __________________^
98  | |         &assoc_predicate,
99  | |         &object_terms,
100 | |         true,
...   |
104 | |         limit,
105 | |     );
    | |_____^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `RSS`

For more information about this error, try `rustc --explain E0596`.
warning: `semsimian_server` (lib) generated 2 warnings
error: could not compile `semsimian_server` (lib) due to previous error; 2 warnings emitted
```

**Comment by justaddcoffee** [comment]
Try this, it seems to work on my end:

```
use std::sync::Mutex;

lazy_static! {
    static ref RSS: Mutex<RustSemsimian> = Mutex::new(get_rss_instance());
}
```

then when you go to use RSS:

```
RSS.lock().unwrap().associations_search(
// args
)
```

**Comment by hrshdhgd** [comment]
~Sorry, I do not follow. Could you direct us to the code file in the repo?~

@justaddcoffee 👍🏽

**Comment by glass-ships** [comment]
ahh nice one, thanks @justaddcoffee !!
`.lock().unwrap()` was the missing key after wrapping RSS in a Mutex!

**Comment by kevinschaper** [comment]
It's not a pretty solution, but to avoid compare unnecessarily waiting for the lock, I kept a non-mutexed instance of RSS around as well. There must have been some extra cloned data hanging around before, because at around 3.5gb, it's using about the same memory now that it was before - which means that even though it feels a bit wrong - I think we can go with this.

**Review by glass-ships — APPROVED** [review]
Thanks for your help, rust semsim team! 

We'd like to make `limit` an optional query param but ran into some trouble, so we're just gonna have it be required by semsim server for now, i think this should work for immediate usage!

**Pull request merged**